src/cuda/api/memory.hpp

```diff
@@ -221,6 +222,117 @@ inline void copy(void *destination, const void *source, size_t num_bytes)
 	}
 }
 
+/**
+ * Synchronously copies data between memory spaces and CUDA arrays.
+ *
+ * @param destination A CUDA array @ref cuda::array::array_t
+ * @param source A pointer to a a memory region of size `destination.size() * sizeof(T)`
+ */
+template<typename T>
+void copy(const array::array_t<T, 3>& destination, const void *source)
+{
+	cudaMemcpy3DParms copyParams = {0};
+
+	// make_cudaPitchedPtr expects a void* pointer
+	void* source_ = const_cast<void*>(source);
+
+	// How to create a pitched ptr from a `const void*` ?
+	copyParams.srcPtr = make_cudaPitchedPtr(
+	    source_, destination.dims()[0] * sizeof(T),  destination.dims()[0], destination.dims()[1]);
+
+	copyParams.dstArray = destination.get();
+
+	cudaExtent ext = make_cudaExtent(destination.dims()[0], destination.dims()[1], destination.dims()[2]);
+
+	copyParams.extent = ext;
+
+	copyParams.kind = cudaMemcpyDefault;
+
+	auto result = cudaMemcpy3D(&copyParams);
+	if (is_failure(result)) {
+		std::string error_message("Synchronously copying into array");
+		throw_if_error(result, error_message);
+	}
+}
+
+/**
+ * Synchronously copies data between memory spaces and CUDA arrays.
+ *
+ * @param destination A pointer to a a memory region of size `source.size() * sizeof(T)`
+ * @param source A CUDA array @ref cuda::array::array_t
+ */
+template<typename T>
+void copy(void* destination, const array::array_t<T, 3>& source)
+{
+	cudaMemcpy3DParms copyParams = {0};
+
+	copyParams.dstPtr = make_cudaPitchedPtr(
+	    destination, source.dims()[0] * sizeof(T),  source.dims()[0], source.dims()[1]);
+
+	copyParams.srcArray = source.get();
+
+	cudaExtent ext = make_cudaExtent(source.dims()[0], source.dims()[1], source.dims()[2]);
+
+	copyParams.extent = ext;
+
+	copyParams.kind = cudaMemcpyDefault;
+
+	auto result = cudaMemcpy3D(&copyParams);
+	if (is_failure(result)) {
+		std::string error_message("Synchronously copying from array");
+		throw_if_error(result, error_message);
+	}
+}
+
+/**
+ * Synchronously copies data between memory spaces and CUDA arrays.
+ *
+ * @param destination A CUDA array @ref cuda::array::array_t
+ * @param source A pointer to a a memory region of size `destination.size() * sizeof(T)`
+ */
+template<typename T>
+void copy(const array::array_t<T, 2>& destination, const void *source)
+{
+	// Consider the padded array:
+	// 
+	// x x x x o o o
+	// x x x x o o o
+	// x x x x o o o
+	// 
+	// o = padded element
+	// x = actually used element of the array
+	// 
+	// The pitch in the example above is 7 * sizeof(T)
+	// The width is 4 * sizeof(T)
+	// The height is 3
```

Comment:
Ok, if I need to read this comment here, then we've not done a good enough job of exposing the concept of padding, and of pitch-width-height.

In fact, do we even have a wrapper for pitched allocation? I don't think so... don't we need it, if we work with pitched pointers? Also, is there never any interaction between pitched allocations and arrays?